### PR TITLE
Fix challenge and implement history challenge screen logic

### PR DESCRIPTION
**Pull request by samymessouak**
# Pull Request: Fix Challenge Logic and Synchronize History

## Description

This pull request addresses several updates and enhancements to the challenge feature and user stats tracking in the Signify application. Below is a detailed summary of the changes:

### 1. **String Update for Challenge History Screen**
   - Added necessary strings to support the redesigned Challenge History Screen.

### 2. **Challenge Model Enhancements**
   - Introduced a `winner` field to the challenge model.
   - Implemented the `updateWinner` function to determine and store the winner of a challenge.

### 3. **User Model Enhancements**
   - Modified the user model to include fields for challenge statistics:
     - Challenges created.
     - Challenges completed.
     - Challenges won.
   - Added functions to retrieve and manipulate these fields.

### 4. **User Model Tests**
   - Developed extensive tests for the newly added functions in the user model to maintain robustness.

### 5. **Game Logic Update**
   - Updated the game logic to allow gameplay for both players, ensuring an engaging and fair experience for all users.

### 6. **Challenge Completion Logic**
   - Enhanced logic to:
     - Remove a completed challenge from the ongoing challenges list.
     - Update user stats accordingly.
   - Added tests to ensure these updates function as expected.

### 7. **Challenge History Screen Redesign**
   - Completely reimplemented the `ChallengeHistoryScreen` class to:
     - Display past challenges with detailed stats for the user.
     - Present a user-friendly interface for browsing past challenges and viewing personal statistics.

### 8. **MainActivity Update**
   - Updated `MainActivity` to align with the redesigned Challenge History Screen.

---

## Time

This feature was developed and tested during **Sprint 8** and **Sprint 9** of the project timeline.

---

Please review the changes, and feel free to provide feedback. We aim to ensure this update improves the overall functionality and user experience of the application.

### CODE REVIEW

File: app/src/main/java/com/github/se/signify/model/common/user/FirestoreUserRepository.kt [hunk 1]
```diff
@@ -530,4 +532,78 @@ class FirestoreUserRepository(
...
+      value: Any,
+      onSuccess: () -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    val userRef = db.collection(collectionPath).document(userId)
+    userRef
+        .update(fieldName, value)
+        .addOnSuccessListener { onSuccess() }
+        .addOnFailureListener { e -> onFailure(e) }
+  }
```
Comment (nada-ezziti) [inline app/src/main/java/com/github/se/signify/model/common/user/FirestoreUserRepository.kt:585]: I don't really understand the utility of incrementField, it's the same logic as updateUserField. Additionally the function do not provide any Success or Failure cases unlike updateUserField. So I think you can just remove it and use the second one with value = 1.

File: app/src/main/java/com/github/se/signify/model/common/user/FirestoreUserRepository.kt [hunk 2]
```diff
@@ -530,4 +532,78 @@ class FirestoreUserRepository(
...
+
+  override suspend fun getChallengesCreated(userId: String): Int {
+    return getUserField(userId, "challengesCreated")
+  }
+
+  override suspend fun getChallengesWon(userId: String): Int {
+    return getUserField(userId, "challengesWon")
+  }
+
```
Comment (nada-ezziti) [inline app/src/main/java/com/github/se/signify/model/common/user/FirestoreUserRepository.kt:598]: You should add onSuccess and onFailure parameters to be consistent with the other functions of the Repository and to ensure the security of the application.

**Review by nada-ezziti — CHANGES_REQUESTED** [review]
I reviewed the pull request and noticed some improvements.  You can see my comments below.

Additionally,  the code coverage should be increased to address the gaps highlighted by SonarCloud, ensuring a more robust and reliable codebase.

File: app/src/main/java/com/github/se/signify/model/common/user/FirestoreUserRepository.kt (hunk 2 above)
Comment (samymessouak) [reply app/src/main/java/com/github/se/signify/model/common/user/FirestoreUserRepository.kt]: Nice catch! I did it.

File: app/src/main/java/com/github/se/signify/model/common/user/FirestoreUserRepository.kt (hunk 1 above)
Comment (samymessouak) [reply app/src/main/java/com/github/se/signify/model/common/user/FirestoreUserRepository.kt]: Yes, I should have use only the function in the UserViewModel, did it!

File: app/src/main/java/com/github/se/signify/model/common/user/UserRepository.kt [hunk 3]
```diff
@@ -111,4 +111,28 @@ interface UserRepository {
...
 
   fun getStreak(userId: String, onSuccess: (Long) -> Unit, onFailure: (Exception) -> Unit)
+
+  fun addPastChallenge(userId: String, challengeId: String)
+
+  fun getPastChallenges(
+      userId: String,
+      onSuccess: (List<Challenge>) -> Unit,
+      onFailure: (Exception) -> Unit
+  )
```
[the remark this answers is not in the document]
Comment (samymessouak) [reply app/src/main/java/com/github/se/signify/model/common/user/UserRepository.kt:121]: As thomas did, we will refactor this for next sprint, i already opened a task for this.

[the remark this answers is not in the document]
Comment (nada-ezziti) [reply app/src/main/java/com/github/se/signify/model/common/user/UserRepository.kt:121]: Okey I saw you opened a PR for that : https://github.com/Signify-epfl/signify-app/issues/436

**Comment by samymessouak** [comment]
I also have a task concerning the code coverage, i'm working on it.

**Comment by nada-ezziti** [comment]
> I also have a task concerning the code coverage, i'm working on it.

Since you're going to address the coverage in your "in development" task : https://github.com/Signify-epfl/signify-app/issues/347  and the meeting is approaching I think we can merge this PR .

**Review by nada-ezziti — APPROVED** [review]

**Pull request merged**